server/src/main/java/com/ncc/neon/services/ClusterService.java

```diff
@@ -23,16 +26,190 @@
+                                                   List<Map<String, Object>> data) {
+        Map<String, Object> datum = data.get(0);
+        Map<String, Object> extraKeySets = new HashMap<>();
+        for (String key : datum.keySet()) {
+            if (!key.equals(aggregationNameKey) && !key.equals(fieldNameKey)) {
+                extraKeySets.put(key, new HashSet<>());
+            }
+        }
+        return extraKeySets;
+    }
+
+    /**
+     * Traverses the original data and puts it in the correct bins in the new data.
+     *
+     * @param fieldNameKey the key that corresponds with the field name
+     * @param aggregationNameKey the key that corresponds with the overall counts of each bin
+     * @param data the original result data
+     * @param newData the newly clustered data
+     * @param extraKeySets the
+     */
+    private void aggregateNumbersInNewData(String fieldNameKey, String aggregationNameKey, List<Map<String,
+            Object>> data, List<Map<String, Object>> newData, Map<String, Object> extraKeySets) {
+        Iterator<Map<String, Object>> newDataIter = newData.iterator();
+        int oldDataIndex = 0;
+        while (newDataIter.hasNext()) {
+            // reset the extra keys for each new bin
+            for (String key : extraKeySets.keySet()) {
+                extraKeySets.put(key, new HashSet<>());
+            }
+
+            // determine the boundaries and aggregated count for this bin
+            BigDecimal currAgg = new BigDecimal("0");
+            Map currNewBin = newDataIter.next();
+            ArrayList alo = (ArrayList) currNewBin.get(fieldNameKey);
+            BigDecimal start = (BigDecimal) alo.get(0);
+            BigDecimal end = (BigDecimal) alo.get(1);
+
+            // traverse the old data a total of one time
+            while (oldDataIndex < data.size()) {
+                // retrieve the old data's field name
+                BigDecimal oldBinValue = new BigDecimal(data.get(oldDataIndex).get(fieldNameKey).toString());
+
+                // if old >= new start && old <= new end
+                if (oldBinValue.compareTo(start) != -1
+                        && oldBinValue.compareTo(end) != 1) {
+
+                    // check extra keys
+                    for (String key : extraKeySets.keySet()) {
+                        ((Set) extraKeySets.get(key)).add(data.get(oldDataIndex).get(key));
+                    }
+
+                    // accumulate aggregate numbers
+                    BigDecimal oldAgg = new BigDecimal(data.get(oldDataIndex).get(aggregationNameKey).toString());
+                    currAgg = currAgg.add(oldAgg);
+
+                    // move to the next data in the old results
+                    oldDataIndex++;
+                } else if (oldBinValue.compareTo(end) == 1) { // old > new end i.e. old bin moved past this curr new bin
+                    break;
```

Comment:
Great catch.  At first I was thinking "I hope he's not looping through the whole results twice!" but I see that you've already thought of that :)